cds-snc/cppd-medical-report-sails
# Physicians sign their declaration on the medical report

# What this does
Adds the Declaration/signature page for Medical Practitioners.

Added a bunch of useful VIRTUAL columns on the MedicalReport model, these should be useful in other contexts - the MA View should be updated to use them, and we can use them in other views around.

## To test
- Fill out an entire medical report, patient and doctor portions
- When the dashboard is showing all sections complete, click Submit to Service Canada
- See the declaration and signature screen
- Try the edit links
- Sign your name, or just a scribble
- Click submit, and see our Confirmation screen!

## Note on Sequelize getterMethods vs. VIRTUAL columns
When I added the getter methods on the MedicalReport model, some model updates started breaking. I'm unclear on if this is a bug or not, but I found [this github issue](https://github.com/sequelize/sequelize/issues/10555) that seems to describe what I'm seeing.  The following describes what I found in my testing, and the changes I made.

When updating a model with `getterMethods`, updating the model with a where clause would break. Ie:

```js
await MedicalReport.update({
    // property assignments
}, {
    where: {
        applicationCode: req.session.applicationCode
    }
});
```

The above update would throw errors on the getterMethods because the properties they were accessing were undefined.

To fix this issue, I found that by querying the model first, then updating it, would fix the issue:

```js
let medicalReport = await MedicalReport.findOne({
    where: {
        applicationCode: req.session.applicationCode
    },
});
// then later update
await medicalReport.update({
    // property assignments
});
```

There were also a few other circumstances where getterMethods were unavailable in certain contexts, and these situations were resolved by switching the getterMethods to VIRTUAL columns on the Sequelize model.

Instead of:

```js
    getterMethods: {
      // ...
      stopWorkingText: function () {
        if (this.stopWorking !== null) {
          return sails.__(stopWorking[this.getDataValue(stopWorking)]);
        }
      },
    },
```

Use Virtual column instead:

```js
    // ...
    stopWorking: Sequelize.INTEGER, // fk to support
    stopWorkingText: {
      type: Sequelize.VIRTUAL,
      get() {
        if (this.stopWorking !== null) {
          return sails.__(stopWorking[this.stopWorking]);
        }
      }
    },
```
